Processing recommendations
==========================

Currently, Qiita supports the processing of raw data from:

#. Target gene barcoded sequencing
#. Shotgun sequencing

Note that the selected processing are mainly guided so we can perform meta-analyses, this is combine different studies,
even from different wet lab techniques or sequencing technologies.


Target gene barcoded sequencing
-------------------------------

For this you can start with raw, pre-demultiplexing data or per-sample FASTQ. Either way, you will need to
"Split libraries and QC", which uses the default in QIIME 1.9.1. Once your demultiplexed and QCed artifact is created
you need to select which processing to perform. There are two main ideologies/methodologies to process target
gene data: sequence clustering and sequence deblur.

Sequencing deblur (preferred)
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

For this we use `deblur <https://github.com/biocore/deblur>`_. Here 2 BIOM tables are generated by default: fina.biom and final.only-16s.biom. The former is the full biom table, which can be used with any target gene and wetlab work;
the latter is the trimmed version to those sequences that match Greengenes at 80% similarity, a really basic and naive filtering. Each of those BIOM tables, is accompanied by a FASTA that contains
the representative sequences. The OTU IDs are given by the unique sequence.

Note that deblur needs all sequences to be trimmed at the same length, thus the recommended pipeline is to trim everything at 150bp and the deblur.

Sequencing clustering
^^^^^^^^^^^^^^^^^^^^^

Here we use close reference picking, for an explanation of the different picking methods see
`"Subsampled open-reference clustering creates consistent, comprehensive OTU definitions and scales to billions of sequences" <https://peerj.com/articles/545/>`_.
Here we generate a single BIOM table with the OTUs/per-sample. The OTU IDs are given based on the reference database selected.

Currently, we have the reference databases: Greengenes version 3_8-97, Silva 119 and Unite 7. Depending on your selection is if the reference has a phylogenetic tree.

Shotgun sequencing
------------------

Qiita currently has one shotgun metagenomics data analysis pipeline: `Shogun <https://msystems.asm.org/content/3/6/e00069-18>`_.

The current workflow is as follows:

#. Removal of adapter sequence and quality control: `Atropos <https://github.com/jdidion/atropos/>`_
#. Removal of host contamination using `Bowtie2 <http://bowtie-bio.sourceforge.net/bowtie2/index.shtml>`_
#. Taxonomy profiling using choice of three different aligners and two different reference databases; see sections below

Note that we recommend only uploading sequences that have already been through QC and human sequence removal. However, we
recommend that all sequence files go through adapter and quality control within the system to ensure they are ready for
subsequent analyses. Currently, the command removes adaptor sequences (only KAPA HyperPlus with iTru, which are compatible
with Illumina TruSeq).

For host removal we currently support *Danio Rerio* (zebrafish), *Drosophila Melanogaster* (fruit fly), *Mus Musculus* (mouse),
*Rattus Norvegicus* (rat), and Enterobacteria phage phiX174 (the Illumina spike-in control).

Note that the Shogun command produces 4 output artifacts:
- The Alignment Profile BIOM artifact, which contains the alignment files
- A Taxonomic Prediction - phylum BIOM artifact, which contains the taxonomic predictions based on the alignment
- A Taxonomic Prediction - genus BIOM artifact, which contains the taxonomic predictions based on the alignment
- A Taxonomic Prediction - species BIOM artifact, which contains the taxonomic predictions based on the alignment
The 3 Taxonomic Prediction files can be used for subsequent analysis and visualization.

Shogun aligners
^^^^^^^^^^^^^^^

#. Bowtie2: The classical ultrafast short sequence aligner. Based on-FM indexing of genome sequences to achieve
   efficient memory and CPU performance. We tuned the parameter setting for Bowtie2 to achieve optimal
   alignment accuracy for typical shotgun metagenome datasets.

   - Version: 2.3.5.1
   - Alignment file format: SAM
   - Website: http://bowtie-bio.sourceforge.net/bowtie2/index.shtml
   - Citation: Langmead B, Salzberg S. Fast gapped-read alignment with Bowtie 2. Nature Methods. 2012, 9:357-359.

#. BURST: A novel sequence aligner featuring **exhaustive** (in contrast to heuristic) alignment against the entire
   reference genome database, hence achieving the highest accuracy, though with relatively long running time.

   - Version: 0.99.8
   - Alignment file format: b6o (BLAST tabular output, i.e., `-outfmt 6`)
   - Website: https://github.com/knights-lab/BURST
   - Citation: Al-Ghalith, Gabriel and Dan Knights. BURST enables optimal exhaustive DNA alignment for big data. DOI 2017:doi.org/10.5281/zenodo.806850
   - Note: Manuscript under review.

#. UTree
   A sequence classifier based on _k_-mer signature matching along a tree structure. Analogous to Kraken but with higher computational efficiency. The fastest option.

   - Version: 2.0 RF
   - Alignment file format: custom mapping file
   - Website: https://github.com/knights-lab/UTree
   - Citation: Al-Ghalith, Gabriel and Dan Knights. Faster and lower-memory metagenomic profiling with UTree. DOI: 10.5281/zenodo.998252

Shogun reference databases
^^^^^^^^^^^^^^^^^^^^^^^^^^

#. WoLr1 ("Web of Life" release 1): An even representation of microbial diversity, selected using an prototype
   selection algorithm based on the MinHash distance matrix among all non-redundant bacterial and archaeal genomes
   from NCBI (RefSeq and GenBank, complete and draft), plus several genome quality control criteria. A
   high-quality reference phylogeny is available for this genome pool, enabling subsequent
   phylogeny-based analyses. Also available are curated taxonomic annotations, based on NCBI and GTDB
   systems.

   - Domains: Bacteria, Archaea
   - Number of genomes: 10,575
   - Total length (bp): 32,861,886,373
   - Citation: Zhu Q, Mai U, Pfeiffer W, et al. Phylogenomics of 10,575 genomes reveals evolutionary
     proximity between domains Bacteria and Archaea. Nat Commun. 2019. 10(1):5477. doi: 10.1038/s41467-019-13443-4.
   - Numbers of taxonomic units:

     - Kingdoms: 2
     - Phyla: 146
     - Classes: 89
     - Orders: 196
     - Families: 422
     - Genera: 2,081
     - Species: 9,105
     - Strains: 89
     - Note: Nucleotide sequences per genome were concatenated with a linker of 20 "N"s.

#. Rep94: NCBI representative and reference microbial genomes, corresponding to RefSeq release 94.

   - Domains: Bacteria, Archaea
   - Number of genomes: 5,808
   - Total length (bp): 23,165,526,011
   - Note: Nucleotide sequences per genome were concatenated with a linker of 20 "N"s.
   - Numbers of taxonomic units:

     - Kingdoms: 2
     - Phyla: 38
     - Classes: 85
     - Orders: 186
     - Families: 427
     - Genera: 1,931
     - Species: 5,636
     - Strains: 84

#. Rep82: NCBI representative and reference microbial genomes, corresponding to RefSeq release 82.

   - Not available anymore for new processing
   - Domains: Bacteria, Archaea, Viruses/Viroids
   - Number of genomes: 10,519
   - Total length (bp): 20,387,349,319
   - Note: Plasmids were isolated from bacterial and archaeal host genomes and considered as separate genomes.
   - Numbers of taxonomic units:

     - Kingdoms: 6
     - Phyla: 55
     - Classes: 362
     - Orders: 182
     - Families: 452
     - Genera: 2,264
     - Species: 11,852
     - Strains: 4,263

Metatranscriptome sample processing
------------------------------------

Sample processing guidelines for metatranscriptomic (metaT) data
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Total community RNA extracted from samples contain both coding and non-coding RNA. Typically, ribosomal RNA make up >90% of the library if not depleted prior to library construction. Ribosomal depletion allows for mRNA enrichment. Even if you are dealing with ribosomal RNA subtracted cDNA libraries, there will be some
residual ribosomal RNA in the libraries that you want to remove/separate from the non ribosomal RNA sequences.

Ribosomal read filtering
^^^^^^^^^^^^^^^^^^^^^^^^

`SortMeRNA <https://bioinfo.lifl.fr/RNA/sortmerna/>`_
is used for removal of ribosomal reads from quality filtered metaT data

Latest SortMeRNA version: v2.1

Input: Quality filtered metaT reads (FASTA/FASTQ) 
Ribosomal reads are identified by searching against pre-curated rRNA databases. Currently, rRNA databases covering bacteria, archaea and eukarya were downloaded and indexed from `SILVA <https://www.arb-silva.de>`_ and `Rfam <https://rfam.xfam.org>`_.
Currently indexed databases and their clustering ids:

- silva-bacterial-16s-id 90%
- silva-bacterial-23s-id 98%
- silva-archaeal-16s-id 95%
- silva-archaeal-23s-id 98%
- silva-eukarya-18s-id 95%
- silva-eukarya-28s-id 98%
- rfam-5s-database-id 98%
- rfam-5.8s-database-id 98%

The above databases and ID cut-offs were chosen to work with a range of samples including more diverse/complex environmental samples.

Building Custom databases
^^^^^^^^^^^^^^^^^^^^^^^^^
Custom databases can also be built in addition to the above mentioned databases.
Custom databases can be built by using the using the `ARB package <https://www.arb-silva.de/download/arb-files/>`_ to extract FASTA files for:

- 16S bacteria, 16S archaea and 18S eukarya using SSURef_NR99_119_SILVA_14_07_14_opt.arb
- 23S bacteria, 23S archaea and 28S eukarya using LSURef_119_SILVA_15_07_14_opt.arb

The built databases will then have to be indexed before running SortMeRNA. 
Reference database(s) and their corresponding indexes separated by "," and multiple databases are separated by ":"


SortMeRNA Usage
^^^^^^^^^^^^^^^
SortMeRNA filters the ribosomal from the non-ribosomal reads from the input sample dataset (via BLAST search)and outputs two fasta/q files containing the ribosomal and non-ribosomal reads respectively. 
Additionally, a summary file showing the proportion of reads matching to each of the screened ribosomal databases can also be made available. 
Default options have been set to report only the best alignment per read reaching E-value. 
For non ribo-depleted samples (i.e. total RNA), the ribosomal reads obtained from SortMeRNA can be further used in taxonomic/compositional analysis. 
In the case of ribo-depleted samples, only the non-ribosomal reads are used in downstream analyses such as assembly, mapping, differential gene abundance analyses etc.
